### ISSUE TITLE
Reduce the hardcoded knowledge of kernel structure in bootstrapper

### ISSUE THREAD

**Issue by janvrany** [powerlang/powerlang#47]
There is just soo much knowledge about the structure of Kernel in bootstrapper, see all of these:

```
DMRBootstrapper >> #undermethods
DMRBootstrapper >> #optimizedMethods
VirtualSmalltalkRuntime class >> #initialize
SmalltalkInterperter class >> #initialize
MethodEvaluator class >> #initialize
VirtualSmalltalkRuntime >> #initializeEmulationOverrides
```

If you look inside there's a lot of overlap - is it intentional or just copy-paste error?

Consider the emulation overrides (`#initializeEmulationOverrides`). You debug the metaphysics emulator and look what weird method it is emulating, then look into Bee Kernel and there's no such method? We have kernel, but in fact it is half described in `bee-dmr` repo and half in `powerlang` repo and what's where is unclear and there's no information that (large) part of it is actually in `powerlang` repo in completely different files.

---

This should all be described in the kernel itself and bootstrapper should extract this from the kernel if needed. 

Also - if simulation requires traditional primitives, they should also go to kernel, not in `#initializeEmulationOverrides`. Besides, this can be useful / required for traditional runtime and/or semi-hosting.

**Comment by shingarov** [comment]
Remnants of Smalltalk-80 where there are two types, Immediate and Pointer, and the VM knows the exact internal representation of both (Pointer being very complex and inflexible) and there is no way to extend that.  So I am afraid this is opening a very big can of worms.

**Comment by janvrany** [comment]
@shingarov Yes, but that's on a different level, that's object model / runtime and how much of that runtime knowledge should go to kernel is unclear (that's the problem of DMR, these things are creeping into base library making them harder to change). But this is not what I'm talking about here.

What I'm talking is - for example - do we really need to manually list all undermessages ([whatever that means](#46)) in `VirtualSmalltalkRuntime` and then (again?) in `MethodEvaluator`? What is the list in `DMRBootstrapper >> #optimizedMethods` good for ([some explanation would be great](#49)). Do we need all that at all?

There's more to it. To build certain objects (such as method dictionaries) the code uses simulation (see `ImageSegmentBuilder`) presumably to make bootstrap independent on an actual implementation method dictionaries. Yet, for some other parts we just build it by hand, hardcoding the structure of kernel into the bootstrap (done - surprisingly - not in segment builder but in `VirtualSmalltalkImage`, see method `#createSpecies:` for instance). We have mixture of both approaches.

I do understand there has to be some coupling (at least in forseeable future), but now I feel there's too much. If I change the
kernel (like, remove metaclasses, use a new intrinsic, use something else than symbols as selectors or change the lookup logic) - what else I have to update in bootstrapper in order to make it work? We should make the list as short as possible and clear (that is, documented)

**Comment by melkyades** [comment]
@janvrany I agree with most of your diagnose, so what do you suggest?

One last comment:

> Also - if simulation requires traditional primitives, they should also go to kernel, not in `#initializeEmulationOverrides`. Besides, this can be useful / required for traditional runtime and/or semi-hosting.

I don't see the reason, the kernel doesn't need them as primitives.

**Comment by janvrany** [comment]
> so what do you suggest?
I do not know what I'm talking about now, but let me try. 

The base idea is to mark special methods in the kernel itself:
  * for underprimitives, add each of them to kernel's `ProtoObject` (anywhere, really) with, say, `<underprim>` annotation and a comment describing what is should do. Bootstrapper then can read the kernel and scan it for list of underprims (and check there are no duplicates and that the backend support all of them). This prevents for accidental mess-ups. This is also helpful when one reads the kernel code, changes it and implements new backend (for new arch)
  * for undermethods, pretty much the same. 
  * for primitives - pretty much the same. Kernel needs them, in order to be bootstrapped. Non-DMR runtime may need them too. They may be useful for semi-hosting too, if you want to run kernel without having everything implemented. 
 
Then there're some other special things, `invokedMethods` and `optimizedMethods` - what is difference to undermethods is not clear to me but the code suggests this is yet another concept.

**Comment by melkyades** [comment]
For undermethods, they are already in ProtoObject (mostly?) but an <undermethod> annotation could be useful. It will require improvements to the compiler. 

For underprimitives, they are part of the JIT/AOT compiler, so they don't go to ProtoObject but to InlineMessageLinker. A pragma might be better than what we have now which is a pseudo-pragma (see the symbol on top of the definition of each underprimitive).

For primitives, I think this varies from implementation to implementation. Maybe we can split it into submodules/overlays. Think Kernel/dmr, Kernel/bootstrap, Kernel/omr. Each implentation may fill the gaps in a different way. Even undermethods would go to Kernel/dmr.
